### PR TITLE
[api-extractor] Merge tokens in a separate post-processing method after original construction instead of during construction

### PR DESCRIPTION
**Pull request by zelliott**
## Summary

Fixes https://github.com/microsoft/rushstack/issues/3561 and improves token merging in a few other cases. Also improves the fix for https://github.com/microsoft/rushstack/pull/3354 by accomplishing https://github.com/microsoft/rushstack/pull/3354#discussion_r855678045.

## Details

The `_buildSpan` method is a very complicated method. You provide it excerpt tokens, a `Span`, and a bag of state, and the method goes off and recursively traverses the `Span`, modifying the provided excerpt tokens and state in the process. To further complicate things, the state object changes the behavior of the `_buildSpan` method itself throughout its recursion. Both @mrshllstock and I have found it very difficult to debug and reason through this method.

The method is responsible for building up excerpt tokens and token ranges, and also tries to be intelligent during this construction and merge tokens where possible. We want to merge tokens where possible in order to serialize the smallest possible token object to the `.api.json` file.

After debugging this method for an hour or so in order to determine the appropriate fix for https://github.com/microsoft/rushstack/issues/3561, I was unable to find a fix that didn't feel like more duct tape on top of an already brittle and overly-complicated method (i.e. the fix essentially involved adding more state to the state object being passed around in an attempt to determine whether tokens should be merged together in certain scenarios).

Thus, I decided to pull out the token merging into a separate method that runs after the tokens are initially built. While this is a less performant solution, it is a much easier solution to understand and makes future debugging easier. Doing this not only fixed https://github.com/microsoft/rushstack/issues/3561, but improved token merging in other unexpected cases.

The new `_condenseTokens` runs in worst-case $O(n * \max(n, m))$, where $n$ is the number of excerpt tokens and $m$ is the number of token ranges. The two outer-loops run in a combined $O(n)$ time (despite being nested loops), because each iteration of the inner loop effectively either (1) performs no merging and moves to the next excerpt token in the outer loop or (2) merges and decreases the number of excerpt tokens by at least one. Thus, the two outer loops can effectively be thought of as one loop through the excerpt tokens. Within this loop, each `splice` operation runs in $O(n)$ time, and each token range update runs in $O(m)$ time, this gives us a combined $O(n * \max(n, m))$ runtime.

In the worst case, you have a huge number of content excerpt tokens and few token ranges, resulting in many merges. We might be able to improve the performance of this method with a more advanced algorithm, but I didn't think it was necessarily worth it given the small average number of excerpt tokens per API item.

## How it was tested

Ran rush rebuild and manually examined the generated `.api.json` files. Also added a dedicated test case for the scenario mentioned in  https://github.com/microsoft/rushstack/issues/3561.

### CODE REVIEW

File: apps/api-extractor/src/generators/ExcerptBuilder.ts [hunk 1]
```diff
@@ -214,54 +203,105 @@ export class ExcerptBuilder {
...
-          previousTokenM1.kind === ExcerptTokenKind.Content &&
-          previousTokenM1.text.trim() === '.' &&
-          previousTokenM2.kind === ExcerptTokenKind.Reference
+          prevPrevToken &&
+          prevPrevToken.kind === ExcerptTokenKind.Reference &&
+          prevToken.kind === ExcerptTokenKind.Content &&
+          prevToken.text.trim() === '.' &&
+          currentToken.kind === ExcerptTokenKind.Reference &&
+          !startOrEndIndices.has(currentIndex) &&
+          !startOrEndIndices.has(currentIndex - 1)
```
Comment (zelliott) [inline apps/api-extractor/src/generators/ExcerptBuilder.ts:256]: This "qualified name merging was copied over", it's not new code, but I think we should revisit it (in a subsequent PR). There are a few different scenarios where this merging can occur:

```
export namespace A {
  export class B {}
}

export enum C {
  D = 'D',
}

export class E {
  static f: number;
}

export function example1(): A.B {}
export function example2(): C.D {}
export function example3(): typeof E.f {}
```

There might be more examples as well. But why even merge in these cases? Isn't it more useful to have links to both `A` and `B` from `example1`, not only `B`? @octogonz maybe you can provide some historical context.

Comment (octogonz) [reply apps/api-extractor/src/generators/ExcerptBuilder.ts:256]: I wrote that code on the scramble up the mountain of PR #951. Not sure how much thought was put into it. 😅

File: common/changes/@microsoft/api-extractor/condense-tokens_2022-07-31-14-58.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/api-extractor",
+      "comment": "Condense ExcerptTokens in post-processing step after original construction instead of attempting to do this mid-construction",
```
Comment (octogonz) [inline common/changes/@microsoft/api-extractor/condense-tokens_2022-07-31-14-58.json:5]:
```suggestion
      "comment": "Fix a regression where .api.json excerpts were sometimes missing tokens (GitHub #3561) and generally improve the quality of excerpt generation",
```

Change log messages should describe what problem we solved for end users. They probably aren't interested in how we implemented it. 😉

**Review by octogonz — APPROVED** [review]

**Comment by octogonz** [comment]
Released with API Extractor 7.29.1 - thanks for getting this fixed!

**Pull request merged**